cpp/include/cudf/io/text/trie.hpp

```diff
@@ -0,0 +1,191 @@
+  // 1) merge is_accepting state in to the most significant bit of the
+  // corrosponding transition, and use a mask to access both values. 2) change
+  // layer_offsets to uint8_t, max string length would be 253 2^8-3 (two values
+  // reserved: empty string, and error state)
+ private:
+  rmm::device_uvector<uint16_t> _layer_offsets;
+  rmm::device_uvector<char> _tokens;
+  rmm::device_uvector<uint16_t> _transitions;
+  rmm::device_uvector<uint8_t> _match_length;
+
+ public:
+  trie(rmm::device_uvector<uint16_t>&& layer_offsets,
+       rmm::device_uvector<char>&& tokens,
+       rmm::device_uvector<uint16_t>&& transitions,
+       rmm::device_uvector<uint8_t>&& match_length)
+    : _layer_offsets(std::move(layer_offsets)),
+      _tokens(std::move(tokens)),
+      _transitions(std::move(transitions)),
+      _match_length(std::move(match_length))
+  {
+  }
+
+  static trie create(std::string const& pattern,
+                     rmm::cuda_stream_view stream,
+                     rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+
+  {
+    return create(std::vector<std::string>{pattern}, stream, mr);
+  }
+
+  static trie create(std::vector<std::string> const& patterns,
+                     rmm::cuda_stream_view stream,
+                     rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+  {
+    std::vector<uint16_t> layer_offsets;
+    std::vector<char> tokens;
+    std::vector<uint16_t> transitions;
+    std::vector<uint8_t> match_length;
+
+    // create the trie tree
+    auto root = std::make_unique<trie_builder_node>();
+    for (auto& pattern : patterns) { root->insert(pattern); }
+
+    // flatten
+    auto sum = 1;
+    layer_offsets.emplace_back(0);
+    transitions.emplace_back(sum);
+    match_length.emplace_back(root->match_length);
+
+    auto nodes = std::queue<std::unique_ptr<trie_builder_node>>();
+    nodes.push(std::move(root));
+
+    while (nodes.size()) {
+      layer_offsets.emplace_back(sum);
+      auto layer_size = nodes.size();
+      for (uint32_t i = 0; i < layer_size; i++) {
+        auto node = std::move(nodes.front());
+        nodes.pop();
+        sum += node->children.size();
+        transitions.emplace_back(sum);
+        for (auto& item : node->children) {
+          match_length.emplace_back(item.second->match_length);
+          tokens.emplace_back(item.first);
+          nodes.push(std::move(item.second));
+        }
+      }
+    }
+
+    match_length.emplace_back(false);
+
+    // allocate device memory
+
+    auto device_layer_offsets = rmm::device_uvector<uint16_t>(layer_offsets.size(), stream, mr);
+    auto device_tokens        = rmm::device_uvector<char>(tokens.size(), stream, mr);
+    auto device_transitions   = rmm::device_uvector<uint16_t>(transitions.size(), stream, mr);
+    auto device_match_length  = rmm::device_uvector<uint8_t>(match_length.size(), stream, mr);
```

Comment:
You can use the cool [vector-factories](https://github.com/rapidsai/cudf/blob/fc95992e4a380f756bd76f888bc49e83a7ec2cd3/cpp/include/cudf/detail/utilities/vector_factories.hpp#L125) to convert `std::vector` into `device_uvector`
```suggestion
    auto device_match_length  = cudf::detail::make_device_uvector_async(match_length, stream, mr);
```
This will also copy the data.